include/zephyr/modbus/modbus.h

```diff
@@ -384,6 +395,55 @@ int modbus_iface_get_by_name(const char *iface_name);
 typedef int (*modbus_raw_cb_t)(const int iface, const struct modbus_adu *adu,
 				void *user_data);
 
+/**
+ * @brief Custom function code handler function signature.
+ *
+ * Modbus allows user defined function codes which can be used to extend
+ * the base protocol. These callbacks can also be used to implement
+ * function codes currently not supported by Zephyr's Modbus subsystem.
+ *
+ * If an error occurs during the handling of the request, the handler should
+ * signal this by setting excep_code to a modbus exception code.
+ *
+ * User data pointer can be used to pass state between subsequent calls to
+ * the handler.
+ *
+ * @param iface      Modbus interface index
+ * @param rx_adu     Pointer to the received ADU struct
+ * @param tx_adu     Pointer to the outgoing ADU struct
+ * @param excep_code Pointer to possible exception code
+ * @param user_data  Pointer to user data
+ *
+ * @retval           true If response should be sent, false otherwise
+ */
+typedef bool (*modbus_custom_cb_t)(const int iface,
+				const struct modbus_adu *const rx_adu,
+				struct modbus_adu *const tx_adu,
+				uint8_t *const excep_code,
+				void *const user_data);
+
+/**
+ * @brief Custom function code definition.
+ */
+struct modbus_custom_fc {
+	sys_snode_t node;
+	modbus_custom_cb_t cb;
+	void *user_data;
+	uint8_t fc;
+	uint8_t excep_code;
```

Comment:
Please document these or have them in a `@cond INTERNAL_HIDDEN` section if they are not meant to be API

Reply:
@lindblandro Please guard it "in a @cond INTERNAL_HIDDEN section".

Reply:
Added the guard.